src/newrelic.ts

```diff
@@ -40,53 +60,367 @@ function addSegment(spec: Spec) {
+              tx_id: spec.data.meta.tx,
+              mi_id: spec.data.meta.mi,
+              fullMessage: this.extractFullMessage(spec)
+            }
+            if (spec.ctx.actdef) {
+              metadata.plugin_name = spec.ctx.actdef.plugin_fullname;
+              metadata.pattern = spec.ctx.actdef.pattern;
+            }
+          
+            if (event === 'outward') {
+              metadata.duration = spec.ctx.duration;
+              metadata.endTime = spec.data.meta.end;
+              metadata.res = spec.data.res;
+              metadata.manualEndTime = Date.now();
+            }
+            if (event === 'inward' && !metadata.startTime) {
+              metadata.startTime = spec.data.meta.start;
+              metadata.manualStartTime = Date.now();
+            }
+            return metadata;
+          },
+        updateSpecList(specMetadata: TelemetrySpecMetadata) {
+          const spec = this.specList.find((s) => s.mi_id === specMetadata.mi_id);
+          if (spec) {
+            Object.assign(spec, specMetadata);
+            if (spec.manualEndTime && spec.manualStartTime && !spec.dispatched) {
+              this.sendTracing(spec)
+                .catch(err => console.log(err));
+              spec.dispatched = true;
+            }
+          } else {
+            this.specList.push(specMetadata);
+          }
+        },
+        async sendTracing(spec: TelemetrySpecMetadata): Promise<string|void|boolean> {
+          return new Promise((resolve, reject) => {
+            const tracingSpec = {
+              id: spec.mi_id,
+              'trace.id': spec.tx_id,
+              timestamp: spec.manualStartTime,
+              attributes: {
+                'duration.ms': spec.manualEndTime! - spec.manualStartTime!,
+                plugin_name: spec.plugin_name,
+                pattern: spec.pattern,
+                name: `${spec.plugin_name} ~ ${spec.pattern}`,
+                'parent.id': spec.tx_id,
+                fullMessage: spec.fullMessage,
+              },
+            };
+            const baseData = [
+              {
+                common: {
+                  attributes: {
+                    "service.name": this.SERVICE_NAME,
+                    host: this.HOST,
+                  },
+                },
+                spans: [tracingSpec],
+              }
+            ];
+            const requestData = JSON.stringify(baseData);
+            const options = {
+              hostname: 'trace-api.newrelic.com',
+              port: 443,
+              path: '/trace/v1',
+              method: 'POST',
+              headers: {
+                'Content-Type': 'application/json',
+                'Api-Key': this.API_KEY,
+                'Data-Format': 'newrelic',
+                'Data-Format-Version': '1',
+              },
+            };
+            const httpRequest = request(options, (res: any) => {
+              // res.on('data', (d: any) => console.log(d));
```

Comment:
Let's discuss the following - do we want to handle response errors? If yes, then how?

I understand that it's a fire-and-forget, but as a user of the plugin, if the API key is wrong, I would at least like to be made aware of the problem as soon as possible.

I can think of at least two strategies:
- resolve when the response is received, if the client code wants fire-and-forget - then they simply need not await on the promise, e.g. `/* do not await */ sendTracing(spec)`
- return an event emitter which supports events, such as (but not limited to): `"sent" | "requestError" | "responseError"`

Let me know what you think. This issue could be addressed in a separate PR, if necessary.

Reply:
Maybe the plugin user can provide a callback to handle the error?